server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -646,35 +753,39 @@ private static ShardSearchRequest buildShardSearchRequest(
 
     private static void executeShardTasks(QueryPerNodeState state) {
         int idx;
-        final int totalShardCount = state.searchRequest.shards.size();
+        final NodeQueryRequest nodeQueryRequest = state.searchRequest;
+        var shards = nodeQueryRequest.shards;
+        final int totalShardCount = shards.size();
         while ((idx = state.currentShardIndex.getAndIncrement()) < totalShardCount) {
             final int dataNodeLocalIdx = idx;
             final ListenableFuture<Void> doneFuture = new ListenableFuture<>();
             try {
-                final NodeQueryRequest nodeQueryRequest = state.searchRequest;
                 final SearchRequest searchRequest = nodeQueryRequest.searchRequest;
                 var pitBuilder = searchRequest.pointInTimeBuilder();
-                var shardToQuery = nodeQueryRequest.shards.get(dataNodeLocalIdx);
+                int translatedIndex = state.shardsToQuery.applyAsInt(dataNodeLocalIdx);
+                var shardToQuery = shards.get(translatedIndex);
                 final var shardId = shardToQuery.shardId;
+                ShardSearchRequest r = state.shardSearchRequests == null ? null : state.shardSearchRequests[translatedIndex];
```

Comment:
A bit messy too for the time being, but kind of neat to save building the requst again as a win in this step, we can refine this further to make it look nicer and maybe also dedup some of the lucene internal query object building down the line.